common/remote/hooks/s3/s3file.cpp

```diff
@@ -0,0 +1,721 @@
+    {
+        close();
+    }
+    catch (...)
+    {
+    }
+}
+
+void S3FileWriteIO::close()
+{
+    CriticalBlock block(cs);
+    if (!blobWritten)
+        file->writeBlob(0, nullptr, stats);
+}
+
+offset_t S3FileWriteIO::appendFile(IFile *file, offset_t pos, offset_t len)
+{
+    throwUnexpected();
+    return 0;
+}
+
+size32_t S3FileWriteIO::write(offset_t pos, size32_t len, const void * data)
+{
+    if (len)
+    {
+        CriticalBlock block(cs);
+        //Very strange semantics for a proof of concept - only allow a single write to the file.
+        //A full implementation will need to either
+        //  write to a temporary file, and then copy to the s3 file when the file is closed.
+        //  use the multi-part upload functionality (has a minimum part size of 5Mb)
+        assertex(!blobWritten);
+        file->writeBlob(len, data, stats);
+        blobWritten = true;
+    }
+    return len;
+}
+
+void S3FileWriteIO::setSize(offset_t size)
+{
+    UNIMPLEMENTED;
+}
+
+void S3FileWriteIO::flush()
+{
+}
+
+unsigned __int64 S3FileWriteIO::getStatistic(StatisticKind kind)
+{
+    return stats.getStatistic(kind);
+}
+
+//---------------------------------------------------------------------------------------------------------------------
+
+static Aws::S3::S3Client createAwsClient()
+{
+    //There should be a default region, (and a default client), but allow the region to be overridden with s3:@region//,
+    //in which a new client would be created.
+    // Set up the request
+    Aws::Client::ClientConfiguration configuration;
+    configuration.region = "eu-west-2";
+
+#ifdef FIXED_CREDENTIALS
+    //The following code allows the access id/secret to come from a value that had been saved away in a secrets manager
+    constexpr const char * myAccessKeyId = "<id>";
+    constexpr const char * myAccessKeySecret = "<secret>";
+    auto credentials = std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(Aws::String(myAccessKeyId), Aws::String(myAccessKeySecret));
+    return Aws::S3::S3Client(credentials, configuration);
+#else
+    //Retrieve the details from environment variables/file/current environment
+    return Aws::S3::S3Client(configuration);
+#endif
+}
+
+static Aws::S3::S3Client & getAwsClient()
+{
+    static Aws::S3::S3Client client = createAwsClient();
+    return client;
+}
+
+
+S3File::S3File(const char *_s3FileName) : fullName(_s3FileName)
+{
+    const char * filename = fullName + strlen(s3FilePrefix);
+    const char * slash = strchr(filename, '/');
+    assertex(slash);
+
+    bucketName.append(slash-filename, filename);
+    objectName.set(slash+1);
+}
+
+bool S3File::getTime(CDateTime * createTime, CDateTime * modifiedTime, CDateTime * accessedTime)
+{
+    ensureMetaData();
+    if (createTime)
+        createTime->clear();
+    if (modifiedTime)
+    {
+        modifiedTime->clear();
+        modifiedTime->set((time_t)(modifiedMsTime / 1000));
+    }
+    if (accessedTime)
+        accessedTime->clear();
```

Comment:
I assume create/accessed time isn't available.
If unknown, would it be better to set create time to modified time